src/Functions/parseDateTime.cpp

```diff
@@ -0,0 +1,1747 @@
+                    cur += literal.size();
+                    return cur;
+                }
+            }
+
+            template <typename T, bool check_space = true>
+            static Pos readNumber2(Pos cur, Pos end, T & res)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 2, "readNumber2 requires size >= 2");
+
+                res = (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                return cur;
+            }
+
+            template <typename T, bool check_space = true>
+            static Pos readNumber3(Pos cur, Pos end, T & res)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 3, "readNumber4 requires size >= 3");
+
+                res = (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                return cur;
+            }
+
+            template <typename T, bool check_space = true>
+            static Pos readNumber4(Pos cur, Pos end, T & res)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 4, "readNumber4 requires size >= 4");
+
+                res = (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                return cur;
+            }
+
+            static void ensureSpace(Pos cur, Pos end, size_t len, const String & msg)
+            {
+                if (cur > end || cur + len > end)
+                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Unable to parse because {}", msg);
+            }
+
+            template <bool check_space = true>
+            static Pos assertChar(Pos cur, Pos end, char ch)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 1, "assertChar requires size >= 1");
+
+                if (*cur != ch)
+                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Expect char {}, but {} provided", String(ch, 1), String(*cur, 1));
+
+                ++cur;
+                return cur;
+            }
+
+            static Pos mysqlDayOfWeekTextShort(Pos cur, Pos end, Date & date)
+            {
+                ensureSpace(cur, end, 3, "Parsing DayOfWeekTextShort requires size >= 3");
+
+                String text(cur, 3);
+                Poco::toLowerInPlace(text);
+                auto it = dayOfWeekMap.find(text);
+                if (it == dayOfWeekMap.end())
+                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Unknown day of week short text {}", text);
```

Comment:
`LOGICAL_ERROR`s are data consistency or programming errors, i.e. problems a user should _never_ see. What about `CANNOT_PARSE_TEXT`?